### PR TITLE
Fixed #24733 -- Passed the triggering exception to 40x error handlers

### PR DESCRIPTION
**Pull request by claudep**

### CODE REVIEW

File: django/core/handlers/base.py [hunk 1]
```diff
@@ -80,10 +82,21 @@ def make_view_atomic(self, view):
...
             callback, param_dict = resolver.resolve_error_handler(status_code)
-            response = callback(request, **param_dict)
+            # Unfortunately, inspect.getargspec result is not trustable enough
+            # depending on the callback wrapping in decorators (frequent for handlers).
+            # Falling back on try/except:
+            try:
+                response = callback(request, **dict(param_dict, exception=exception))
+            except TypeError:
+                warnings.warn(
+                    "Error handlers should accept an exception parameter, update "
```
Comment (timgraham) [inline django/core/handlers/base.py]: , -> .

File: django/core/handlers/base.py [hunk 2]
```diff
@@ -162,42 +175,42 @@ def get_response(self, request):
                             % (middleware_method.__self__.__class__.__name__))
                 response = response.render()
 
-        except http.Http404 as e:
+        except http.Http404 as exc:
```
Comment (timgraham) [inline django/core/handlers/base.py]: could you do these renames in a separate commit to keep this patch cleaner?

File: django/views/defaults.py [hunk 3]
```diff
@@ -7,7 +8,7 @@
 # therefore need @requires_csrf_token in case the template needs
 # {% csrf_token %}.
 @requires_csrf_token
-def page_not_found(request, template_name='404.html'):
+def page_not_found(request, exception, template_name='404.html'):
```
Comment (timgraham) [inline django/views/defaults.py]: Was not passing the exception content to the context intentional here as well?

File: docs/ref/views.txt [hunk 4]
```diff
@@ -61,7 +61,12 @@ these with your own custom views, see :ref:`customizing-error-views`.
 The 404 (page not found) view
 -----------------------------
 
-.. function:: defaults.page_not_found(request, template_name='404.html')
+.. function:: defaults.page_not_found(request, exception, template_name='404.html')
+
+.. versionchanged:: 1.9
+
+    The signature of ``page_not_found()`` changed in Django 1.9. The function
```
Comment (timgraham) [inline docs/ref/views.txt]: Since this renders with "Changed in Django 1.9:" as a prefix, I think "change in Django 1.9" is a bit redundant.

File: docs/ref/views.txt [hunk 5]
```diff
@@ -145,6 +164,9 @@ data). If not specifically handled, Django will consider the current request a
 
 ``django.views.defaults.bad_request``, is otherwise very similar to the
 ``server_error`` view, but returns with the status code 400 indicating that
-the error condition was the result of a client operation.
+the error condition was the result of a client operation. By default, nothing
+related to the exception that triggered the view is passed to the template
+context, as the exception message might contain sensible information like
```
Comment (timgraham) [inline docs/ref/views.txt]: sensitive

File: docs/releases/1.9.txt [hunk 6]
```diff
@@ -255,6 +255,9 @@ Requests and Responses
 
 * The debug view now shows details of chained exceptions on Python 3.
 
+* The default 40x error views now accept a second positional parameter, the
+  exception that triggered the view.
+
 Tests
```
Comment (timgraham) [inline docs/releases/1.9.txt]: Please add a deprecation section about view signatures needing to be adapted.

File: django/core/handlers/base.py (hunk 2 above)
Comment (claudep) [inline django/core/handlers/base.py]: Pushed in 69125ce

File: django/views/defaults.py (hunk 3 above)
Comment (claudep) [inline django/views/defaults.py]: I did not do it because of some reticence in previous tickets. But I think we should include it anyway. Our users are grown-up, right :-).

File: docs/ref/views.txt [hunk 7]
```diff
@@ -69,8 +75,9 @@ special view devoted to handling 404 errors. By default, it's the view
 simple "Not Found" message or loads and renders the template ``404.html`` if
 you created it in your root template directory.
 
-The default 404 view will pass one variable to the template: ``request_path``,
-which is the URL that resulted in the error.
+The default 404 view will pass two variables to the template: ``request_path``,
+which is the URL that resulted in the error, and ``exception`` which is the
```
Comment (timgraham) [inline docs/ref/views.txt]: comma after exception

File: docs/ref/views.txt [hunk 8]
```diff
@@ -118,7 +132,9 @@ default, call the view ``django.views.defaults.permission_denied``.
 
 This view loads and renders the template ``403.html`` in your root template
 directory, or if this file does not exist, instead serves the text
-"403 Forbidden", as per :rfc:`2616` (the HTTP 1.1 Specification).
+"403 Forbidden", as per :rfc:`2616` (the HTTP 1.1 Specification). The template
+context contains ``exception`` which is the unicode representation of the
```
Comment (timgraham) [inline docs/ref/views.txt]: comm after exception

File: docs/topics/http/views.txt [hunk 9]
```diff
@@ -141,6 +141,11 @@ but can easily be overridden if you need any custom behavior. Simply specify
 the handlers as seen below in your URLconf (setting them anywhere else will
 have no effect).
 
+.. versionchanged:: 1.9
```
Comment (timgraham) [inline docs/topics/http/views.txt]: IMO, we could omit this note since the signatures aren't discussed here.

File: tests/view_tests/tests/test_debug.py [hunk 10]
```diff
@@ -81,7 +81,7 @@ def test_403(self):
         'OPTIONS': {
             'loaders': [
                 ('django.template.loaders.locmem.Loader', {
-                    '403.html': 'This is a test template for a 403 error.',
+                    '403.html': 'This is a test template for a 403 error ({{ exception }}).',
```
Comment (timgraham) [inline tests/view_tests/tests/test_debug.py]: Add a similar test for 404?

Comment (claudep) [inline tests/view_tests/tests/test_debug.py]: It was a good idea to request a test for 404, because I realized the unicode representation of Resolver404 is not especially interesting to be displayed in custom 404 templates. I added some exception introspection to display something more useful (IMHO).

**Comment by timgraham** [comment]
Not sure I understand why `Resolver404` deserves special treatment. As far as I can tell, a similar thing would happen with the other exceptions if a message isn't provided?
```
>>> str(Resolver404)
"<class 'django.core.urlresolvers.Resolver404'>"
>>> str(PermissionDenied)
"<class 'django.core.exceptions.PermissionDenied'>"
```

**Comment by claudep** [comment]
The problem is that the default representation of Resolver404 is ugly. Something like force_text(exception):
`u"{u'path': u'other_non_existing_url/', u'tried': [[<RegexURLPattern None ^$>], [<RegexURLPattern None ^non_existing_url/>], [<RegexURLPattern None ^server_error/>], [<RegexURLPattern None raises/$>], [<RegexURLPattern None raises400/$>], [<RegexURLPattern None raises403/$>], [<RegexURLPattern None raises404/$>], [<RegexURLPattern None raises500/$>], [<RegexURLPattern my404 technical404/$>], [<RegexURLPattern None classbased404/$>], [<RegexURLResolver <module 'django.conf.urls.i18n' from '/home/claude/virtualenvs/djangogit/local/lib/python2.7/site-packages/django/conf/urls/i18n.pyc'> (None:None) ^i18n/>], [<RegexURLPattern None ^jsi18n/$>], [<RegexURLPattern None ^jsi18n/app1/$>], ...]]}"
`
So I want to get the nice message that would be given by `Http404("This page is no more...")`, but not the ugly one from `Resolver404`.

File: django/views/defaults.py [hunk 11]
```diff
@@ -15,8 +17,21 @@ def page_not_found(request, template_name='404.html'):
     Context:
         request_path
             The path of the requested URL (e.g., '/app/pages/bad_page/')
+        exception
+            Any interesting message from the exception which triggered the 404,
+            or the exception class name
     """
-    context = {'request_path': request.path}
+    try:
```
Comment (timgraham) [inline django/views/defaults.py]: Does this version seem any better? (so we don't have to repeat the first statement two times)
```python
exception_repr = exception.__class__.__name__
try:
    message = exception.args[0]
except (AttributeError, IndexError):
    pass
else:
    if isinstance(message, six.text_type):
        exception_repr = message
```
Also, I'd add some comments to explain the purpose.

File: django/views/defaults.py [hunk 12]
```diff
@@ -15,8 +17,21 @@ def page_not_found(request, template_name='404.html'):
     Context:
         request_path
             The path of the requested URL (e.g., '/app/pages/bad_page/')
+        exception
+            Any interesting message from the exception which triggered the 404,
```
Comment (timgraham) [inline django/views/defaults.py]: The message from the exception (if one was supplied)...

File: docs/ref/views.txt [hunk 13]
```diff
@@ -61,7 +61,13 @@ these with your own custom views, see :ref:`customizing-error-views`.
 The 404 (page not found) view
 -----------------------------
 
-.. function:: defaults.page_not_found(request, template_name='404.html')
+.. function:: defaults.page_not_found(request, exception, template_name='404.html')
+
+.. versionchanged:: 1.9
```
Comment (timgraham) [inline docs/ref/views.txt]: I think these annotations should go at the bottom of the section (but tell me if you've seen other examples of them at the top like this)

**Comment by timgraham** [comment]
Thanks for the explanation. I missed the fact that Resolver404 is initialilized with a dictionary, so a comment about that would be helpful.

**Comment by timgraham** [comment]
LGTM
